pkg/sign/certificate.go

```
@@ -132,7 +136,10 @@ func (f *Fulcio) GetCertificate(ctx context.Context, keypair Keypair, opts *Cert
 	}
 
 	// Sign JWT subject for proof of possession
-	subjectSignature, _, err := keypair.SignData(ctx, []byte(subject))
+	// Fulcio doesn't support verifying Ed25519ph signatures at the moment,
+	// so if we're signing with the prehash variant, we need to adapt the
+	// signer to sign with Ed25519 instead.
+	subjectSignature, _, err := keypair.SignData(context.WithValue(ctx, adaptEd25519Ctx{}, true), []byte(subject))
```

We did something similar in Cosign - https://github.com/sigstore/cosign/blob/eed2a117b8e707d702f43e113bc3488a4e3af457/cmd/cosign/cli/sign/sign.go#L633

---

Just so I am understanding this. If a client uses Ed25519ph when requesting a cert from fulcio, we silently fallback to Ed25519 but **with the same key**?

---

Correct, with the same key. We only fallback to Ed25519 to sign the proof of possession, not the artifact signature (Rekor does properly verify ed25519ph).

The reason Fulcio doesn't support Ed25519ph is because crypto/x509 doesn't support the prehash variant. There's no public OID to represent the key algorithm. When I work on the next major release of Fulcio, I'll see what I can to hack this in.

---

I appreciate the work being done here to try to work around a lack of support for ed25519ph in Fulcio, but I wonder if instead we should return an error for now instead.

What you have here works, but if / when Fulcio supports ed25519ph in the future, figuring out how to unwind this and migrate people off is going to be a bit tricky. Is there a particular need for this use-case right now?

---

The use-case is a bit niche, to support hashedrekord entries with ed25519 signatures and with a Fulcio certificate (note that Rekor already supports verifying both ed25519 and -ph signatures).

Note that if you are signing an entry with a key and _not_ requesting a Fulcio certificate, then this isn't needed. This would only be if you use Cosign with `--issue-certificate` or if you specify the signature algorithm (not yet supported, https://github.com/sigstore/cosign/pull/3497).

I'm OK with an error for now.

---

I've dropped support for ed25519ph with Fulcio. I also added e2e testing when signing with a self-managed key.